src/utils/triton.h

```diff
@@ -212,6 +254,77 @@ class TritonRequestView : public TritonRequestWrapper<TritonRequestView> {
   TRITONBACKEND_Request *handle_ = nullptr;
 };
 
+
+template<class Actual>
+class TritonResponseWrapper {
+ public:
+  OBufferDescr AllocateOutputBuffer(const IOMeta &out_info, device_type_t preferred_device) {
+    auto output_shape = array_shape(out_info.shape);
+    TRITONBACKEND_Output *triton_output;
+    TRITON_CALL(TRITONBACKEND_ResponseOutput(This(), &triton_output, out_info.name.c_str(),
+                                             to_triton(out_info.type), output_shape.data(),
+                                             output_shape.size()));
+    TRITONSERVER_MemoryType memtype = to_triton(preferred_device);
+    int64_t memid = 0;
+    auto t_size = TRITONSERVER_DataTypeByteSize(to_triton(out_info.type));
+    OBufferDescr buffer{};
+    buffer.size = volume(output_shape) * t_size;
+    TRITON_CALL(
+        TRITONBACKEND_OutputBuffer(triton_output, &buffer.data, buffer.size, &memtype, &memid));
+    buffer.device = to_dali(memtype);
+    buffer.device_id = memid;
+    return buffer;
+  }
+
+ private:
+  Actual &This() noexcept {
+    return static_cast<Actual &>(*this);
+  }
+
+  const Actual &This() const noexcept {
+    return static_cast<const Actual &>(*this);
+  }
+};
+
+/** @brief Owning handle for a Triton response. */
+class TritonResponse :
+    public UniqueHandle<TRITONBACKEND_Response *, TritonResponse>,
+    public TritonResponseWrapper<TritonResponse> {
+ public:
+  DALI_INHERIT_UNIQUE_HANDLE(TRITONBACKEND_Response *, TritonResponse)
+
+  static TritonResponse New(TritonRequestView request) {
+    TRITONBACKEND_Response *handle;
+    TRITON_CALL(TRITONBACKEND_ResponseNew(&handle, request));
+    return TritonResponse(handle);
+  }
+
+  static void DestroyHandle(TRITONBACKEND_Response *response) {
+    LOG_IF_ERROR(TRITONBACKEND_ResponseDelete(response),
+                 make_string("Failed releasing a response."));
+  }
```

Comment:
Is it correct? From Triton docs:
```c++
/// Send a response. Calling this function transfers ownership of the
/// response object to Triton. The caller must not access or delete
/// the response object after calling this function.
///
/// \param response The response.
/// \param send_flags Flags associated with the response. \see
/// TRITONSERVER_ResponseCompleteFlag. \see
/// TRITONSERVER_InferenceResponseCompleteFn_t.
/// \param error The TRITONSERVER_Error to send if the response is an
/// error, or nullptr if the response is successful.
/// \return a TRITONSERVER_Error indicating success or failure.
TRITONBACKEND_DECLSPEC TRITONSERVER_Error* TRITONBACKEND_ResponseSend(
    TRITONBACKEND_Response* response, const uint32_t send_flags,
    TRITONSERVER_Error* error);
```

Looks like deleting response after sending it is forbidden

Reply:
That's why SendResponse consumes the TritonResponse. It's even documented ;)

Reply:
Although TritonResponse still needs a destructor if we construct a response and don't send it.